### PR TITLE
[BLOG] 2022 03 07 fine dining for indices

### LINKED ISSUE #711: [BLOG] Fine Dining for Indices
**Describe the blog post your would like to write**

This blog will cover the options available to fine tune an index as well as the API calls necessary to manipulate them. 

**What is the title of the blog post?**

Fine Dining For Indices

**Who are the authors?**

nateboot

**What is the proposed posting date?**

2022-03-07

### PR DESCRIPTION
**Pull request by nateynateynate**
### Description
Blog about Tuning an Index
 
### Issues Resolved
#711 

### Check List
- [X] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the BSD-3-Clause License.

### CODE REVIEW

File: _posts/2022-03-07-fine-dining-for-indices.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+---
+
+## Is Your Cluster Eating Healthy? 
+
+There are a lot of reasons you might want to fine-tune indices on an OpenSearch cluster.  Every workflow has different requirements, and the default behavior of OpenSearch might not suit your use case.  Before ingesting with reckless abandon to fill an index with data, enjoy some insight on tuning an index.  A small bit of foresight can help fine tune a cluster for speed or space efficiency to suit a particular workflow. 
+
+## Are You Going To Eat All of That? 
+
+By default, OpenSearch indexes all the fields in each document it ingests. Enabling or disabling this on a per field basis allows for a customized index with only specific fields in it. This does not nullify the original document - a field that isn’t indexed can still be retrieved and parsed from the `_source` field, but the data is not part of the index, nor can it be aggregated by visualizations.  Consider this dev tool snippet -
```
Comment (stockholmux) [inline _posts/2022-03-07-fine-dining-for-indices.md:20]: 'OpenSearch indexes all the fields in each document it ingests' and 'a field that isn’t indexed' -> the term "index" (and variants) are really over loaded in OpenSearch (it's a noun, verb, etc.). Maybe rephrase for clarity.

File: _posts/2022-03-07-fine-dining-for-indices.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+```
+PUT my_index_name
```
Comment (stockholmux) [inline _posts/2022-03-07-fine-dining-for-indices.md:23]: maybe a better index name and since you're doing multiple `PUT my_index_name` calls, the user following along might have a weird experience when trying to PUT the same index multiple times.

**Review by stockholmux — CHANGES_REQUESTED** [review]
Overall very good.

The only inaccuracy revolves around the 'enabled' bit not working for `text` but it doesn't change the premise of your example (it will need to be tweaked though for sure)

I think the biggest changes that need to happen is to show more full examples - show creating an index with these features, storing a document and retrieving or querying for it. I think that will make the context much clearer.

File: _posts/2022-03-07-fine-dining-for-indices.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+{
+    "mappings": {
+        "properties": {
+            "eat_me": {
+                "type": "text"
+            },
+            "field_that_does_not_need_to_be_indexed": {
+                "type": "text",
+                "enabled": false
```
[the remark this answers is not in the document]
Comment (nateynateynate) [reply _posts/2022-03-07-fine-dining-for-indices.md:32]: Ahh - I think the equivalent of this for non object fields is actually `index` - `index:false` for text, `enabled:false` for objects. I'd like to avoid using an example with an object type so I'm going to modify the examples to use `index: false`

[the remark this answers is not in the document]
Comment (nateynateynate) [reply _posts/2022-03-07-fine-dining-for-indices.md]: I've changed the examples to use type text and `index: false`

File: _posts/2022-03-07-fine-dining-for-indices.md (hunk 1 above)
Comment (nateynateynate) [reply _posts/2022-03-07-fine-dining-for-indices.md]: Let me know if the latest push alleviates your concerns on clarity.

File: _posts/2022-03-07-fine-dining-for-indices.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            },
+            "drink_me": {
+                "type": "text"                
+            }
+        }
+    }
+}
+```
+
+`my_index_name` will refuse to index any ingested fields  called `field_that_does_not_need_to_be_indexed` . If you later decide to retrieve the document you will still find the values stored in the `_source` element, just not in the index. **This tradeoff is one of efficiency.** The size of your index will be somewhat proportional to the amount of data you store in it. Deciding not to index certain fields means a smaller index, which means a smaller cluster.
```
[the remark this answers is not in the document]
Comment (nateynateynate) [reply _posts/2022-03-07-fine-dining-for-indices.md]: I've extended my example to just show that the un-indexed fields can't be searched for.

File: _posts/2022-03-07-fine-dining-for-indices.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+  }
+```
+
+## Save the Leftovers? 
+
+There are also other options for keeping only specific pieces of the `_source` material after indexing. This retains the `_source` document, but not in its entirety.  While this may reduce the size of the resulting `_source` field, you will lose the data you exclude if it were to be re-indexed. It will cease to be. 
+
+
+```
+"mappings": {
```
[the remark this answers is not in the document]
Comment (nateynateynate) [reply _posts/2022-03-07-fine-dining-for-indices.md]: Example provided.

File: _posts/2022-03-07-fine-dining-for-indices.md (hunk 2 above)
Comment (nateynateynate) [reply _posts/2022-03-07-fine-dining-for-indices.md]: Each example uses a different index name so as to compare them near the end.

File: _posts/2022-03-07-fine-dining-for-indices.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+Note the addition of `dynamic: false`. With this option in place, your mapping will never be automatically updated when ingesting a new field.  You can always check if your mapping has changed: 
+
+```
+GET /my_index_name/_mapping
+```
+
+## Can I Keep the Recipe? 
+
+By default the entire document ingested is saved as a field named `_source`. This field can be disabled (it is by default enabled) when creating your index,  usually for storage space efficiency.  Ingested documents are indexed, but the originally submitted document is not stored. In this case, if you decide not to index every field encountered, the data will not be part of the index, nor will it be retrievable from the `_source` field. It will be lost to oblivion.
```
[the remark this answers is not in the document]
Comment (nateynateynate) [reply _posts/2022-03-07-fine-dining-for-indices.md:213]: Done.

File: _posts/2022-03-07-fine-dining-for-indices.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+            },
+            "drink_me": {
+                "type": "text"                
+            }
+        }
+    }
+}
+```
+
+Note the addition of `dynamic: false`. With this option in place, your mapping will never be automatically updated when ingesting a new field.  You can always check if your mapping has changed:
```
[the remark this answers is not in the document]
Comment (nateynateynate) [reply _posts/2022-03-07-fine-dining-for-indices.md]: Reworded accordingly.

File: _posts/2022-03-07-fine-dining-for-indices.md [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+        }
+      }
+    }
+  }
+}
+```
+
+## Can I Keep the Recipe? 
+
+By default the entire document ingested is saved as a field named `_source`. This field can be disabled (it is by default enabled) when creating your index,  usually for storage space efficiency.  Ingested documents are indexed, but the originally submitted document is not stored. In this case, if you decide not to index every field encountered, the data will not be part of the index, nor will it be retrievable from the `_source` field.
```
Comment (stockholmux) [inline _posts/2022-03-07-fine-dining-for-indices.md:213]: On second read, I'm a little iffy calling `_source` a field. It's an object.

So: 
'ingested is saved as a field named `_source`. This field can be disabled' -> 'ingested to the `_source` object. `_source` can be disabled'

and

'nor will it be retrievable from the `_source` field. ' -> 'nor will it be retrievable from the `_source` object.'

File: _posts/2022-03-07-fine-dining-for-indices.md [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+    }
+  }
+}
+```
+
+## Can I Keep the Recipe? 
+
+By default the entire document ingested is saved as a field named `_source`. This field can be disabled (it is by default enabled) when creating your index,  usually for storage space efficiency.  Ingested documents are indexed, but the originally submitted document is not stored. In this case, if you decide not to index every field encountered, the data will not be part of the index, nor will it be retrievable from the `_source` field. 
+
+Consider this index creation API call that would disable saving the `_source` field:
```
Comment (stockholmux) [inline _posts/2022-03-07-fine-dining-for-indices.md:215]: 'saving the `_source` field' -> 'saving the `_source` object'

File: _posts/2022-03-07-fine-dining-for-indices.md [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+```
+
+
+## Flavor To Taste, and Bon Apetít!
+
+A lot of index functionality is controlled by these small, sometimes overlooked stanzas of configuration. Storing the source document, deciding whether to index or not index fields, or even whether new fields should be added on the fly are all things that have a lasting impact on the way it behaves. The way that OpenSearch comes out of the box is suitable for a lot of tasks, but there’s no replacement for having intimacy and forethought with your own data.  Getting comfortable with the index configuration options will help ensure your cluster is ingesting properly. 
+
+## How Was The Service Today?
+
+As a community its important to know whether we’re advocating for all of the users. Please get involved and let us know if we’re providing what is needed to make the most out of OpenSearch. Whether via the [GitHub Project](https://github.com/opensearch-project), the [community forums](https://discuss.opendistrocommunity.dev/) or any of the other ways you can [connect with the community.](https://opensearch.org/connect.html)
```
Comment (stockholmux) [inline _posts/2022-03-07-fine-dining-for-indices.md:398]: 'As a community it's important to know whether we’re' -> 'As a community its important to know the team is'

'Please get involved and let us know if we’re providing what is needed to make the most out of OpenSearch' -> 'Please get involved and speak up if you find anything that would help you get more out of OpenSearch.'

File: _posts/2022-03-07-fine-dining-for-indices.md [hunk 11]
```diff
@@ -0,0 +1,408 @@
+---
+layout: post
+title:  "Fine Dining for Indices"
+authors:
+- nateboot
+
+date: 2022-03-07
```
Comment (stockholmux) [inline _posts/2022-03-07-fine-dining-for-indices.md:7]: change the date to march 9. (Also the file name)

**Review by stockholmux — CHANGES_REQUESTED** [review]
Straight forward changes. Read to publish afterwards.

File: _posts/2022-03-07-fine-dining-for-indices.md [hunk 12]
```diff
@@ -0,0 +1,408 @@
...
+            }
+          }
+        }
+      }
+    }
+  }
+}
+```
+
+Wait a second - ingesting a field that wasn't explicitly mapped added a new mapping. `some_other_field` wasn't there in the mapping until we just happened to post it with a new document. This exemplifies a great dining tip:
```
[the remark this answers is not in the document]
Comment (nateynateynate) [reply _posts/2022-03-07-fine-dining-for-indices.md:144]: Thank you. This is a firmware update for my brain I'll have to work on.

**Review by stockholmux — APPROVED** [review]
LGTM
